### PR TITLE
(api): Support k-nearest neighbor (kNN) search

### LINKED ISSUE #381: Add support for knn
Add support for K Nearest Neighbor in ZIO-elasticsearch.

Updates to the api should allow for programatically defining these fields:

- query_vector
- knn
- similarity

### PR DESCRIPTION
**Pull request by drmarjanovic**
Resolves #381

### CODE REVIEW

File: modules/library/src/main/scala/zio/elasticsearch/executor/HttpExecutor.scala [hunk 1]
```diff
@@ -372,6 +373,27 @@ private[elasticsearch] final class HttpExecutor private (esConfig: ElasticConfig
...
+    sendRequestWithCustomResponse[SearchWithAggregationsResponse](
+      baseRequest
+        .post(uri)
+        .response(asJson[SearchWithAggregationsResponse])
+        .contentType(ApplicationJson)
+        .body(r.toJson)
+    ).flatMap { response =>
+      response.code match {
+        case HttpOk =>
+          println(response.body)
```
Comment (arnoldlacko) [inline modules/library/src/main/scala/zio/elasticsearch/executor/HttpExecutor.scala:390]: Probably forgot to remove this :)

File: modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala [hunk 2]
```diff
@@ -31,6 +31,10 @@ sealed trait ElasticQuery[-S] { self =>
   private[elasticsearch] def toJson(fieldPath: Option[String]): Json
 }
 
+sealed trait ElasticKNNQuery[-S] { self =>
```
Comment (arnoldlacko) [inline modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala:34]: Do we need the `Elastic` prefix, shouldn't it be just `KNNQuery`?

File: modules/library/src/main/scala/zio/elasticsearch/executor/HttpExecutor.scala [hunk 3]
```diff
@@ -372,6 +373,27 @@ private[elasticsearch] final class HttpExecutor private (esConfig: ElasticConfig
...
+          ZIO.succeed(new KNNSearchResult())
```
Comment (arnoldlacko) [inline modules/library/src/main/scala/zio/elasticsearch/executor/HttpExecutor.scala:391]: I just realised when I saw this that this PR is marked as Draft :)

**Review by arnoldlacko — APPROVED** [review]

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala [hunk 4]
```diff
@@ -546,6 +546,46 @@ object ElasticQuery {
...
+   * @param numCandidates
+   *   number of nearest neighbor candidates to consider per shard
+   * @param queryVector
+   *   query vector
+   * @tparam S
+   *   document for which field query is executed
+   * @return
+   *   an instance of [[zio.elasticsearch.query.KNNQuery]] that represents the kNN query to be performed.
+   */
+  final def kNN[S](field: Field[S, _], k: Int, numCandidates: Int, queryVector: Chunk[Double]): KNNQuery[S] =
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:567]: Maybe we could rename `k` to something more meaningful.

File: modules/library/src/test/scala/zio/elasticsearch/domain/TestDocument.scala [hunk 5]
```diff
@@ -13,14 +13,16 @@ final case class TestDocument(
...
+  geoPointField: GeoPoint,
+  vectorField: List[Int]
 )
 
 object TestDocument {
-  implicit val schema
-    : Schema.CaseClass7[String, List[TestSubDocument], LocalDate, Int, Double, Boolean, GeoPoint, TestDocument] =
+  implicit val schema: Schema.CaseClass8[String, List[TestSubDocument], LocalDate, Int, Double, Boolean, GeoPoint, List[
+    Int
+  ], TestDocument] =
```
Comment (dbulaja98) [inline modules/library/src/test/scala/zio/elasticsearch/domain/TestDocument.scala:23]: Can this be formatted better?

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala (hunk 4 above)
Comment (drmarjanovic) [reply modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:567]: I guess `k` makes more sense to users using it. Additionally, it's a kNN search.

**Pull request merged**